nixos/modules/services/network-filesystems/glusterfs.nix

```diff
@@ -40,7 +89,14 @@ in
 
     services.rpcbind.enable = true;
 
+    environment.etc = mkIf (cfg.tlsSettings != null) {
+      "ssl/glusterfs.pem".source = cfg.tlsSettings.tlsPem;
+      "ssl/glusterfs.key".source = cfg.tlsSettings.tlsKey;
+      "ssl/glusterfs.ca".source = cfg.tlsSettings.caCert;
```

Comment:
OK, here I'm not quite sure what `source` exactly does.

Does it just point the symlink `/etc/ssl/glusterfs.key -> theGivenPath`? Or does it take a local path to a file, put it into the nix store, and create a symlink `/etc/ssl/glusterfs.key -> /nix/store/thefile`?

The latter would be bad, because that would make the private key world readable in `/nix/store`.

We definitely need the former, so that e.g. in the case of nixops we can obtain a symlink like `/etc/ssl/glusterfs.key -> /var/run/keys/...` where the target directory is the keys directory managed by nixops and not part of the store.

Can anybody shed light on this?

Reply:
On my machine it just creates the files end up in the nix store :(

```
lrwxrwxrwx 1 root root 28 Sep 17 17:33 glusterfs.ca -> /etc/static/ssl/glusterfs.ca
lrwxrwxrwx 1 root root 29 Sep 17 17:33 glusterfs.key -> /etc/static/ssl/glusterfs.key
lrwxrwxrwx 1 root root 29 Sep 17 17:33 glusterfs.pem -> /etc/static/ssl/glusterfs.pem
```

where:

```
lrwxrwxrwx 1 root root 50 Jan  1  1970 glusterfs.ca -> /nix/store/n9h1w46qjcbsrqpr38i7n6vdqb8cfj16-ca.pem
lrwxrwxrwx 1 root root 61 Jan  1  1970 glusterfs.key -> /nix/store/8skzrsdyx46kjzpz8m5kvj6y9x4qhqyy-cleopatra-key.pem
lrwxrwxrwx 1 root root 57 Jan  1  1970 glusterfs.pem -> /nix/store/54y38wlzipf142mxzki2qh3sz4hp4zxz-cleopatra.pem
```

I'm not sure how to do this properly?

Reply:
Is there a particular reason for requiring the use of `environment.etc`? Compared to, say, using `preStart` or a dedicated unit for this purpose?

Reply:
The original entry I have in my `configuration.nix`:

```
  services.glusterfs = {
    enable = true;
    tlsSettings = {
      tlsKey = /home/pascal/ca/cleopatra-key.pem;
      tlsPem = /home/pascal/ca/cleopatra.pem;
      caCert = /home/pascal/ca/ca.pem; 
    };
  };
```

Reply:
What if you enclose the paths in quotes, so they're treated as string literals and not paths (which will be added to the store on eval)?

Reply:
(My brief reading of the etc builder code suggests it'll take whatever you give it, so long as it looks like a filepath).